src/Interpreters/Cache/FileSegment.cpp

```diff
@@ -324,11 +338,14 @@ void FileSegment::write(const char * from, size_t size, size_t offset)
 
     try
     {
-        cache_writer->write(from, size);
+        /// if `from` is nullptr, then we just allocate and hold space by current segment and it was (or would) be written outside
+        if (cache_writer && from != nullptr)
```

Comment:
I've updated this code, and now `wrtie` and `reserve` are different methods. From the user side, it's almost the same (because passing nullptr directly was not intended and was hidden in `reserve`). But yes, the solution with this nullptr internally looked like a workaround, so I split methods.

Reply:
> Why cannot we write actual data here?

Right now, `NativeWrier` is responsible for writing. Besides writing data, we also need to perform serialization, compression, buffering, etc. We have several layers of abstractions incompatible with the cache interface. We can refactor some code and make some adapters to make it seamless. Or implement some adapter (e.g., implement `WriteBuffer` for file segment or make `FileSegment` accept writer or buffer, we still need to consider how it would look like). But it would require numerous changes. We can make it step by step, though, if we decide it's necessary.